File: frontend/src/api/jsw-stock/list.ts
/**
 * `GET /jsw-stock` — backend-driven paginated JSW Stock list.
 *
 * All filtering, sorting, and pagination is server-driven — do NOT apply any
 * client-side transforms to the returned rows.
 *
 * `buildQuery` strips `undefined` and `""` (and we never forward `null`), so
 * every optional filter is forwarded only when truthy.
 */

import { buildQuery, getList } from "@/api/client"
import type { PaginatedResult } from "@/types/api/envelope"
import type { JswStock, JswStockListQuery } from "@/types/jsw-stock/stock"

/**
 * Normalise `JswStockListQuery` into a plain record safe for `buildQuery`.
 * Keys are snake_case to match the backend `JswStockListQuery` exactly.
 */
function buildParams(
  q: JswStockListQuery,
): Record<string, string | number | undefined> {
  return {
    page: q.page,
    limit: q.limit,
    sortBy: q.sortBy,
    sortOrder: q.sortOrder,
    // single report-date filter ("dd-MM-yyyy", exact match on report_date)
    ...(q.date             ? { date: q.date }                             : {}),
    // 5 per-field exact-match filters — only include when non-empty
    ...(q.party_code       ? { party_code: q.party_code }                 : {}),
    ...(q.sales_order_type ? { sales_order_type: q.sales_order_type }     : {}),
    ...(q.customer_name    ? { customer_name: q.customer_name }           : {}),
    ...(q.sales_office     ? { sales_office: q.sales_office }             : {}),
    ...(q.nco_declared     ? { nco_declared: q.nco_declared }             : {}),
  }
}

/**
 * Fetch a paginated, server-filtered page of JSW Stock rows.
 *
 * @param query - Optional query parameters; defaults to page 1, limit 20,
 *                sorted by `created_at` descending with no active filters.
 * @returns A paginated result containing the matching `JswStock` rows
 *          and standard `PaginationMeta` (total, page, limit, totalPages).
 */
export function listJswStock(
  query: JswStockListQuery = {},
): Promise<PaginatedResult<JswStock>> {
  return getList<JswStock>(`/jsw-stock${buildQuery(buildParams(query))}`)
}

File: backend/app/services/jsw_stock/list.py
# app/services/jsw_stock/list.py
"""JswStock domain: backend-driven paginated listing.

Mirrors services/customer_code/list.py exactly.
No region lookup needed — customer_name is stored denormalised on every document
(set at ingest time from the CustomerCode collection). All filtering and sorting
run in the DB layer; no Python-side post-filtering (backend-api-standards).
"""
from __future__ import annotations

from math import ceil

from ...core.responses import PaginationMeta
from ...models.customer_code import CustomerCode
from ...models.jsw_stock import JswStock
from ...schemas.jsw_stock import JswStockListQuery
from ...schemas.jsw_stock_record import JswStockPublic
from ...utils.jsw_stock.query import build_jsw_stock_filter, build_sort
from .serialize import to_jsw_stock_public


async def list_jsw_stock(
    query: JswStockListQuery,
) -> tuple[list[JswStockPublic], PaginationMeta]:
    """Return a page of JswStock rows plus stable pagination metadata.

    Pipeline: build filter → count → sort/skip/limit → serialize → build meta.
    Every operation runs in MongoDB; no client-side filtering ever occurs.

    Args:
        query: Validated list-query DTO carrying page, limit, sortBy, sortOrder,
               the single ``date`` filter, and the 4 per-field filters
               (sales_order_type, customer_name, sales_office, nco_declared).

    Returns:
        A 2-tuple of (items, meta) where items is the serialised page and meta
        carries pagination bookkeeping consumed by the frontend table component.
    """
    filt = build_jsw_stock_filter(query)

    if query.region:
        region_docs = await CustomerCode.find({"region_id": query.region}).to_list()
        region_codes = [str(doc.id) for doc in region_docs]
        filt["customer_code_id"] = {"$in": region_codes}

    total = await JswStock.find(filt).count()

    docs = (
        await JswStock.find(filt)
        .sort(build_sort(query.sortBy, query.sortOrder))
        .skip(query.skip)
        .limit(query.limit)
        .to_list()
    )

    items = [to_jsw_stock_public(doc) for doc in docs]

    meta = PaginationMeta(
        page=query.page,
        limit=query.limit,
        total=total,
        totalPages=ceil(total / query.limit) if query.limit else 0,
        sortBy=query.sortBy,
        sortOrder=query.sortOrder,
    )

    return items, meta
